crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -979,6 +1051,272 @@ impl MultiProofTask {
             .filter(|proof| !proof.is_empty())
     }
 
+    /// Processes a multiproof message, batching consecutive same-type messages.
+    ///
+    /// Drains queued messages of the same type and merges them into one batch before processing.
+    /// This reduces redundant trie traversals when messages arrive faster than proofs complete.
+    ///
+    /// If a different message type is encountered, it's stored in `pending_msg` for the next
+    /// iteration (preserving order without sending back to channel).
+    ///
+    /// Returns `true` if done, `false` to continue.
+    #[allow(clippy::too_many_arguments)]
+    fn process_multiproof_message(
+        &mut self,
+        msg: MultiProofMessage,
+        pending_msg: &mut Option<MultiProofMessage>,
```

Comment:
It'd be nice to pull out a lot (or all) of these `&mut`s into a separate struct, like `MultiproofBatchCtx` or something. It seems like a lot are just for metrics, if so maybe two structs, `MultiproofBatchCtx` and `MultiproofBatchMetrics`.